omisego-sdk/src/main/java/co/omisego/omisego/websocket/channel/SocketChannel.kt

```diff
@@ -0,0 +1,186 @@
+import co.omisego.omisego.model.socket.SocketTopic
+import co.omisego.omisego.websocket.SocketChannelListener
+import co.omisego.omisego.websocket.SocketClientContract
+import co.omisego.omisego.websocket.SocketConnectionListener
+import co.omisego.omisego.websocket.SocketCustomEventListener
+import co.omisego.omisego.websocket.channel.SocketChannelContract.Dispatcher
+import co.omisego.omisego.websocket.channel.SocketChannelContract.SocketClient
+import co.omisego.omisego.websocket.channel.dispatcher.SocketDispatcherContract
+import co.omisego.omisego.websocket.enum.SocketEventSend
+import co.omisego.omisego.websocket.enum.SocketStatusCode
+import okhttp3.WebSocketListener
+
+/**
+ * A SocketChannel is responsible for handling join or leave from the socket channel.
+ *
+ * @param socketDispatcher A [Dispatcher] is responsible dispatch all events to the client.
+ * @param socketClient A [SocketClient] for sending a message to the eWallet web socket API and close the web socket connection.
+ */
+internal class SocketChannel(
+    override val socketDispatcher: SocketChannelContract.Dispatcher,
+    override val socketClient: SocketChannelContract.SocketClient
+) : SocketClientContract.Channel, SocketChannelContract.Channel, SocketDispatcherContract.SocketChannel {
+    private val channelSet: MutableSet<String> by lazy { mutableSetOf<String>() }
+
+    /**
+     * A [SocketMessageRef] is responsible for create unique ref value to be included in the [SocketSend] request.
+     */
+    override val socketMessageRef: SocketChannelContract.MessageRef = SocketMessageRef().apply { scheme = SocketMessageRef.SCHEME_JOIN }
+
+    /**
+     * An interval of milliseconds for scheduling the interval event such as the heartbeat event which used for keeping the connection alive.
+     * Default 5,000 milliseconds.
+     */
+    override var period: Long = 5000
+
+    /**
+     * Send [SocketEventSend.JOIN] event to the server. Do nothing if the channel has already joined.
+     *
+     * @param topic Join the channel by the given topic.
+     * @param payload (Optional) the additional data you might want to send bundled with the request.
+     */
+    override fun join(topic: String, payload: Map<String, Any>) {
+        if (!joined(topic)) {
+            socketClient.send(createJoinMessage(topic, payload))
+        }
+    }
+
+    /**
+     * Send [SocketEventSend.LEAVE] event to the server. Do nothing if the channel has already left.
+     *
+     * @param topic Leave from the channel by the given topic.
+     * @param payload (Optional) payload you want to send along with the request/.
+     */
+    override fun leave(topic: String, payload: Map<String, Any>) {
+        if (joined(topic)) {
+            socketClient.send(createLeaveMessage(topic, payload))
+        }
+    }
+
+    /**
+     * Send leave event for all currently active channels.
+     */
+    override fun leaveAll() {
+        for (channel in channelSet) {
```

Comment:
Also dangerous, you could get a concurrent access exception, if calling this and you receive a join event in another thread at the same time, you should use a copy of the list (maybe a `toList` or `toIterator` would be enough, not sure which one is creating a new collection)

Reply:
Fixed it